### PR TITLE
descriptor: Use `DescriptorError` instead of `Error` when reasonable

### PR DESCRIPTION
**Pull request by afilini**
Change the return type of the `descriptor!()` macro and `ToWalletDescriptor` to avoid having to map errors.

Also introduce more checks to validate descriptors built using the macro.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature
* [x] I've updated `CHANGELOG.md`

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing

### CODE REVIEW

File: src/descriptor/dsl.rs [hunk 1]
```diff
@@ -317,10 +354,10 @@ macro_rules! apply_modifier {
 ///
 /// let b_items = vec![
 ///     bdk::fragment!(pk(my_key_1))?,
-///     bdk::fragment!(s:pk(my_key_2))?,
-///     bdk::fragment!(s:d:v:older(my_timelock))?,
+///     bdk::fragment!(s: pk(my_key_2))?,
+///     bdk::fragment!(s: d: v: older(my_timelock))?,
```
Comment (RCasatta) [inline src/descriptor/dsl.rs]: spaces inserted on modifier by rustfmt doesn't look good, maybe adding a `#[rustfmt(skip)]` is slightly less ugly

same on `policy.rs:1154`

Comment (afilini) [reply src/descriptor/dsl.rs]: Yeah, makes sense

File: src/descriptor/template.rs [hunk 2]
```diff
@@ -44,24 +45,48 @@ pub type DescriptorTemplateOut = (ExtendedDescriptor, KeyMap, ValidNetworks);
 /// Since [`ToWalletDescriptor`] is implemented for any [`DescriptorTemplate`], they can also be
 /// passed directly to the [`Wallet`](crate::Wallet) constructor.
 ///
-/// ## Example
+/// ## Examples
 ///
 /// ```
```
Comment (LLFourn) [inline src/descriptor/template.rs:50]: ⛏️   When there are multiple examples it is desirable to have a small amount of text before each example explaining what they are intended to teach.

File: src/descriptor/template.rs [hunk 3]
```diff
@@ -44,24 +45,48 @@ pub type DescriptorTemplateOut = (ExtendedDescriptor, KeyMap, ValidNetworks);
...
+///     pk_b: K,
+///     timelock: u32,
+/// }
+///
+/// impl<K: ToDescriptorKey<Segwitv0>> DescriptorTemplate for TimeDecayingMultisig<K> {
+///     fn build(self) -> Result<DescriptorTemplateOut, DescriptorError> {
+///         let mut uncompressed_pk = bdk::bitcoin::PrivateKey::from_wif("L5EZftvrYaSudiozVRzTqLcHLNDoVn7H5HSfM9BAN6tMJX8oTWz6").unwrap();
+///         uncompressed_pk.compressed = false;
+///
+///         // Ok(bdk::descriptor!(wsh(thresh(2,pk(self.pk_a),s:pk(self.pk_b),s:d:v:older(self.timelock))))?)
```
Comment (LLFourn) [inline src/descriptor/template.rs]: ❓  Is this comment meant to be here? I am not sure what it's getting across.

**Review by LLFourn — APPROVED** [review]

File: src/descriptor/template.rs (hunk 3 above)
Comment (afilini) [reply src/descriptor/template.rs]: Nope, looks like I forgot to remove it...

File: src/descriptor/template.rs (hunk 2 above)
Comment (afilini) [reply src/descriptor/template.rs:50]: Makes sense, I'll add some more details

File: src/descriptor/template.rs (hunk 3 above)
Comment (afilini) [reply src/descriptor/template.rs]: Actually, this new test I've added doesn't really make sense: I wanted to show that it correctly fails if you use uncompressed pks in a segwit script, but it's not a compile-time error, it's a runtime error. I would have to not only declare the template but also build a wallet and use it in there to actually show the failure..

Besides, I've also already added a test for that in `src/descriptor/dsl.rs`, so I think I'm just gonna remove this one

**Pull request merged**